Modules/SaltStackConfig/SaltStackConfig.psm1

```diff
@@ -0,0 +1,304 @@
+   Created by:	Brian Wuchner
+   Date:		November 27, 2021
+   Blog:		www.enterpriseadmins.org
+   Twitter:		@bwuch
+  ===========================================================================
+  .SYNOPSIS
+    This function will clear the global variable used to connect to the vRealize Automation SaltStack Config API
+  .DESCRIPTION
+    This function clears a previously created cookie/header used to connect to SaltStack Config
+  .EXAMPLE
+    PS C:\> Disconnect-SscServer
+#>
+  if ($global:DefaultSscConnection) {
+    $global:DefaultSscConnection = $null 
+  } else {
+    write-warning "Not connected to any SaltStack Config servers."
+  } # end if
+}
+
+Function Get-SscData {
+<#
+  .NOTES
+  ===========================================================================
+   Created by:	Brian Wuchner
+   Date:		November 27, 2021
+   Blog:		www.enterpriseadmins.org
+   Twitter:		@bwuch
+  ===========================================================================
+  .SYNOPSIS
+    This function will pass resource/method/arguments to the vRealize Automation SaltStack Config API.
+    It depends on a global variable created by Connect-SscServer.
+  .DESCRIPTION
+    Use this function to call the SaltStack Config API.
+    Additional helper functions will call this function, this is where the majority of the logic will happen.
+  .EXAMPLE
+    PS C:\> Get-SscData -Resource 'minions' -Method 'get_minion_cache'
+#>
+  param(
+    [Parameter(Mandatory=$true)][string]$resource,
+    [Parameter(Mandatory=$true)][string]$method,
+    [System.Collections.Hashtable]$kwarg
+  )
+
+  if (!$global:DefaultSscConnection) {
+    write-warning "Not connected to any SaltStack Config servers."
+    return;
+  } # end if
+
+  if (!$kwarg) {
+    $body = "{`"resource`": `"$resource`", `"method`": `"$method`"}"
+  } else {
+    $body = "{`"resource`": `"$resource`", `"method`": `"$method`", `"kwarg`": $(ConvertTo-Json $kwarg) }"
+  }
+
+  try{
+    $output = Invoke-WebRequest -WebSession $global:DefaultSscConnection.SscWebSession -Method POST -Uri "https://$($global:DefaultSscConnection.SscServer)/rpc" -body $body -ContentType 'application/json'
+    return (ConvertFrom-Json $output.Content)
+  } catch {
+    write-warning $_.Exception.Message
+  }
+}
+
+
+# Lets include a couple sample/helper functions wrappers
+Function Get-SscMaster {
+<#
+  .NOTES
+  ===========================================================================
+   Created by:	Brian Wuchner
+   Date:		November 27, 2021
+   Blog:		www.enterpriseadmins.org
+   Twitter:		@bwuch
+  ===========================================================================
+  .SYNOPSIS
+    This wrapper function will call Get-SscData master.get_master_grains.
+  .DESCRIPTION
+    This wrapper function will return grain details about the SaltStack Config master node.
+  .EXAMPLE
+    PS C:\> Get-SscMaster
+#>
+
+  param(
+    [ValidateSet('RAW','Results')][string]$Return='RAW'

```

Comment:
The json output that is returned includes some additional bits of info, like errors if you have a syntax problem, an riq, etc.  In testing I had a need to look at json, but I believe the most common output need will be the results object in the 'ret' property.  I'm going to set that as the default return value on the existing wrapper functions.  Here are some example return members if this is helpful for confirmation.  Please let me know if there is a better way to handle this.

RAW output (now called Json):
```
Get-SscMinion |Get-Member

   TypeName: System.Management.Automation.PSCustomObject

Name        MemberType   Definition
----        ----------   ----------
Equals      Method       bool Equals(System.Object obj)
GetHashCode Method       int GetHashCode()
GetType     Method       type GetType()
ToString    Method       string ToString()
error       NoteProperty object error=null
ret         NoteProperty System.Management.Automation.PSCustomObject ret=@{count=9; results=System.Object[]}
riq         NoteProperty long riq=19814806475293
warnings    NoteProperty Object[] warnings=System.Object[]
```

Results (now returned by default):
```
Get-SscMinion -Return:Results |Get-Member

   TypeName: System.Management.Automation.PSCustomObject

Name        MemberType   Definition
----        ----------   ----------
Equals      Method       bool Equals(System.Object obj)
GetHashCode Method       int GetHashCode()
GetType     Method       type GetType()
ToString    Method       string ToString()
grains      NoteProperty System.Management.Automation.PSCustomObject grains=@{id=dr-control-01.lab.enterpriseadmins....
key_state   NoteProperty string key_state=accepted
last_seen   NoteProperty string last_seen=2021-10-28T00:09:49.754927
master      NoteProperty string master=salt
minion      NoteProperty string minion=dr-control-01.lab.enterpriseadmins.org
status      NoteProperty string status=lost
```

Reply:
I see. So what I would do is to remove the output parameter and just add some additional processing. If we have warnings returned just write them with Write-Warning before you return the objects. If there are objects returned in ret return them after you have processed the warnings. If there are no objects, but there are errors write them with Write-Error. My assumption here is that there is no case in which there are errors and results at the same time.

Reply:
That makes sense and could reduce the effort to create/maintain the wrapper functions, as all of this could be handled in the proxy Get-SscData function.  I'll get working on this after the next commit.

Reply:
This will be implemented in the next commit.  Marking as resolved.